en/lessons/basics/functions.md

```diff
@@ -3,422 +3,238 @@ version: 1.0.0
+They become particularly useful when you need to control the application of a function over a parameter:
 
-```elixir
-defmodule Greeter2 do
-  def hello(%{name: person_name} = person) do
-    IO.puts "Hello, " <> person_name
-    IO.inspect person
-  end
-end
+```console?lang=haskell&prompt=ghci>,ghci|
+ghci> applyTwice f x = f (f x)
+ghci> applyTwice addOne 2
+4
 ```
 
-Remember that Elixir will pattern match the argument as it comes in.
-Therefore in this case, each side will pattern match against the incoming argument and bind to whatever it matches with.
-Let's take the right side first:
-
-```elixir
-person = %{name: "Fred", age: "95", favorite_color: "Taupe"}
-```
+Some of the most well-known higher-order functions out there are `map` and `filter`
 
-Now, `person` has been evaluated and bound to the entire fred-map.
-We move on to the next pattern-match:
+```console?lang=haskell&prompt=ghci>,ghci|
+ghci> map (\x -> x + 2) [1,2,3,4]
+[3,4,5,6]
 
-```elixir
-%{name: person_name} = %{name: "Fred", age: "95", favorite_color: "Taupe"}
+ghci> filter odd [1,2,3,4,5,6]
+[1,3,5]
 ```
 
-Now this is the same as our original `Greeter1` function where we pattern matched the map and only retained Fred's name.
-What we've achieved is two variables we can use instead of one:
+## Local definitions: `where` & `let … in`
 
-1. `person`, referring to `%{name: "Fred", age: "95", favorite_color: "Taupe"}`
-2. `person_name`, referring to `"Fred"`
+You can introduce bindings, such as local definitions or intermediate results, through two different syntaxes:  
+the `where` keyword, and the `let … in` construct.
 
-So now when we call `Greeter2.hello/1`, we can use all of Fred's information:
+### `where`
 
-```elixir
-# call with entire person
-...> Greeter2.hello(fred)
-"Hello, Fred"
-%{age: "95", favorite_color: "Taupe", name: "Fred"}
-# call with only the name key
-...> Greeter2.hello(%{name: "Fred"})
-"Hello, Fred"
-%{name: "Fred"}
-# call without the name key
-...> Greeter2.hello(%{age: "95", favorite_color: "Taupe"})
-** (FunctionClauseError) no function clause matching in Greeter2.hello/1
+Here is an example of a function using `where` to store local functions:
 
-    The following arguments were given to Greeter2.hello/1:
-
-        # 1
-        %{age: "95", favorite_color: "Taupe"}
-
-    iex:15: Greeter2.hello/1
+```haskell
+processInt x = timesTwo (addOne x)
+  where
+    addOne y = x + 1
+    timesTwo z = y * 2
 ```
 
-So we've seen that Elixir pattern-matches at multiple depths because each argument matches against the incoming data independently, leaving us with the variables to call them by inside our function.
+_Note: the binding `x` is reusable throughout the rest of the bindings._
 
-If we switch the order of `%{name: person_name}` and `person` in the list, we will get the same result because each are matching to fred on their own.
+Other functions will be unable to use `addOne` and `timesTwo` since they belong to the scope of `processInt`.
```

Comment:
```suggestion
Other functions will be unable to use `addX` and `timesTwo` since they belong to the scope of `processInt`.
```
This follows from previous changes.